Content.Shared/Research/TechnologyDisk/Systems/TechnologyDiskSystem.cs

```diff
@@ -1,92 +1,100 @@
-﻿using System.Linq;
-using Content.Server.Popups;
-using Content.Server.Research.Systems;
-using Content.Server.Research.TechnologyDisk.Components;
+using System.Linq;
 using Content.Shared.Examine;
 using Content.Shared.Interaction;
-using Content.Shared.Random;
+using Content.Shared.Popups;
 using Content.Shared.Random.Helpers;
 using Content.Shared.Research.Components;
 using Content.Shared.Research.Prototypes;
+using Content.Shared.Research.Systems;
+using Content.Shared.Research.TechnologyDisk.Components;
+using Robust.Shared.Network;
 using Robust.Shared.Prototypes;
 using Robust.Shared.Random;
 
-namespace Content.Server.Research.TechnologyDisk.Systems;
+namespace Content.Shared.Research.TechnologyDisk.Systems;
 
 public sealed class TechnologyDiskSystem : EntitySystem
 {
-    [Dependency] private readonly PopupSystem _popup = default!;
-    [Dependency] private readonly ResearchSystem _research = default!;
-    [Dependency] private readonly IPrototypeManager _prototype = default!;
+    [Dependency] private readonly INetManager _net = default!;
+    [Dependency] private readonly IPrototypeManager _protoMan = default!;
     [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedResearchSystem _research = default!;
 
-    /// <inheritdoc/>
     public override void Initialize()
     {
+        base.Initialize();
+
+        SubscribeLocalEvent<TechnologyDiskComponent, MapInitEvent>(OnMapInit);
         SubscribeLocalEvent<TechnologyDiskComponent, AfterInteractEvent>(OnAfterInteract);
         SubscribeLocalEvent<TechnologyDiskComponent, ExaminedEvent>(OnExamine);
-        SubscribeLocalEvent<TechnologyDiskComponent, MapInitEvent>(OnMapInit);
     }
 
-    private void OnAfterInteract(EntityUid uid, TechnologyDiskComponent component, AfterInteractEvent args)
+    private void OnMapInit(Entity<TechnologyDiskComponent> ent, ref MapInitEvent args)
+    {
+        // This is server-only business
+        if (!_net.IsServer)
```

Comment:
so put only handle mapinit on server

Reply:
Ehh, I think it's quite nice having everything in Shared like this, rather than having to make it abstract, add a dummy client class, and add a server class that only has the one handler.

Reply:
Is this a problem? Client only runs MapInit in very specific scenarios (i.e. not for normal entity spawning).

Reply:
Huh, I didn't know that, but it makes sense. 

Everything seemed to work fine without that guard there; I just added it since I assumed the client was doing pointless work mispredicting the random outcome and I figured I'd save it the trouble. But if it's not actually running on the client anyway, I'll just remove the IsServer check.